File: pine-go/internal/runtime/stats.go
package runtime

import (
	"sync"
	"sync/atomic"
	"time"
)

// OpStats holds per-operator cumulative execution statistics.
// All fields are updated atomically and are safe for concurrent access.
type OpStats struct {
	ExecCount      int64 // number of successful executions
	SkipCount      int64 // number of times the operator was skipped
	ErrorCount     int64 // number of errors (fatal)
	TotalDurationNs int64 // cumulative execution time in nanoseconds
	MaxDurationNs   int64 // maximum single-execution time in nanoseconds
}

// Stats accumulates per-operator execution statistics across requests.
// Thread-safe for concurrent updates from multiple Execute calls.
type Stats struct {
	mu    sync.RWMutex
	ops   map[string]*OpStats
	// scheduler-level
	runCount        int64
	peakConcurrency int64
}

// NewStats creates a new Stats accumulator.
func NewStats() *Stats {
	return &Stats{
		ops: make(map[string]*OpStats),
	}
}

// PreInitOperators pre-registers all operator names so they appear in
// Snapshot() from startup, even before any requests are processed.
func (s *Stats) PreInitOperators(names []string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, name := range names {
		if _, ok := s.ops[name]; !ok {
			s.ops[name] = &OpStats{}
		}
	}
}

// getOrCreate returns the OpStats for a given operator, creating it if needed.
func (s *Stats) getOrCreate(name string) *OpStats {
	s.mu.RLock()
	st, ok := s.ops[name]
	s.mu.RUnlock()
	if ok {
		return st
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	// Double-check after acquiring write lock
	if st, ok = s.ops[name]; ok {
		return st
	}
	st = &OpStats{}
	s.ops[name] = st
	return st
}

// RecordExec records a successful operator execution.
func (s *Stats) RecordExec(name string, duration time.Duration) {
	st := s.getOrCreate(name)
	atomic.AddInt64(&st.ExecCount, 1)
	ns := duration.Nanoseconds()
	atomic.AddInt64(&st.TotalDurationNs, ns)

	// CAS loop for max
	for {
		cur := atomic.LoadInt64(&st.MaxDurationNs)
		if ns <= cur {
			break
		}
		if atomic.CompareAndSwapInt64(&st.MaxDurationNs, cur, ns) {
			break
		}
	}
}

// RecordSkip records a skipped operator execution.
func (s *Stats) RecordSkip(name string) {
	st := s.getOrCreate(name)
	atomic.AddInt64(&st.SkipCount, 1)
}

// RecordError records a failed operator execution.
func (s *Stats) RecordError(name string, duration time.Duration) {
	st := s.getOrCreate(name)
	atomic.AddInt64(&st.ErrorCount, 1)
	ns := duration.Nanoseconds()
	atomic.AddInt64(&st.TotalDurationNs, ns)

	for {
		cur := atomic.LoadInt64(&st.MaxDurationNs)
		if ns <= cur {
			break
		}
		if atomic.CompareAndSwapInt64(&st.MaxDurationNs, cur, ns) {
			break
		}
	}
}

// RecordRun records a scheduler run (one per Engine.Execute call).
func (s *Stats) RecordRun() {
	atomic.AddInt64(&s.runCount, 1)
}

// RecordConcurrency updates peak concurrency if n exceeds the current peak.
func (s *Stats) RecordConcurrency(n int64) {
	for {
		cur := atomic.LoadInt64(&s.peakConcurrency)
		if n <= cur {
			break
		}
		if atomic.CompareAndSwapInt64(&s.peakConcurrency, cur, n) {
			break
		}
	}
}

// Snapshot returns a point-in-time copy of all operator statistics.
type OpStatsSnapshot struct {
	ExecCount      int64   `json:"exec_count"`
	SkipCount      int64   `json:"skip_count"`
	ErrorCount     int64   `json:"error_count"`
	TotalDurationNs int64  `json:"total_duration_ns"`
	MaxDurationNs   int64  `json:"max_duration_ns"`
	AvgDurationNs   int64  `json:"avg_duration_ns"`
}

// Snapshot returns a point-in-time copy of all operator statistics.
func (s *Stats) Snapshot() map[string]OpStatsSnapshot {
	s.mu.RLock()
	defer s.mu.RUnlock()

	result := make(map[string]OpStatsSnapshot, len(s.ops))
	for name, st := range s.ops {
		exec := atomic.LoadInt64(&st.ExecCount)
		total := atomic.LoadInt64(&st.TotalDurationNs)
		var avg int64
		if exec > 0 {
			avg = total / exec
		}
		result[name] = OpStatsSnapshot{
			ExecCount:       exec,
			SkipCount:       atomic.LoadInt64(&st.SkipCount),
			ErrorCount:      atomic.LoadInt64(&st.ErrorCount),
			TotalDurationNs: total,
			MaxDurationNs:   atomic.LoadInt64(&st.MaxDurationNs),
			AvgDurationNs:   avg,
		}
	}
	return result
}

// SchedulerStatsSnapshot is a point-in-time copy of scheduler-level statistics.
type SchedulerStatsSnapshot struct {
	RunCount        int64 `json:"run_count"`
	PeakConcurrency int64 `json:"peak_concurrency"`
}

// SchedulerSnapshot returns a point-in-time copy of scheduler statistics.
func (s *Stats) SchedulerSnapshot() SchedulerStatsSnapshot {
	return SchedulerStatsSnapshot{
		RunCount:        atomic.LoadInt64(&s.runCount),
		PeakConcurrency: atomic.LoadInt64(&s.peakConcurrency),
	}
}

File: pine-python/pine/stats.py
"""Internal execution statistics tracking (thread-safe)."""
from __future__ import annotations

import threading
from typing import Any


class _OpStats:
    __slots__ = ("exec_count", "skip_count", "error_count",
                 "total_duration_ns", "max_duration_ns", "_lock")

    def __init__(self):
        self._lock = threading.Lock()
        self.exec_count: int = 0
        self.skip_count: int = 0
        self.error_count: int = 0
        self.total_duration_ns: int = 0
        self.max_duration_ns: int = 0

    def record_exec(self, duration_ns: int):
        with self._lock:
            self.exec_count += 1
            self.total_duration_ns += duration_ns
            if duration_ns > self.max_duration_ns:
                self.max_duration_ns = duration_ns

    def record_skip(self):
        with self._lock:
            self.skip_count += 1

    def record_error(self, duration_ns: int):
        with self._lock:
            self.error_count += 1
            self.total_duration_ns += duration_ns
            if duration_ns > self.max_duration_ns:
                self.max_duration_ns = duration_ns

    def snapshot(self) -> dict[str, Any]:
        with self._lock:
            exec_c = self.exec_count
            total = self.total_duration_ns
            return {
                "exec_count": exec_c,
                "skip_count": self.skip_count,
                "error_count": self.error_count,
                "total_duration_ns": total,
                "max_duration_ns": self.max_duration_ns,
                "avg_duration_ns": total // exec_c if exec_c > 0 else 0,
            }


class _Stats:
    def __init__(self):
        self._lock = threading.Lock()
        self._ops: dict[str, _OpStats] = {}
        self._run_count: int = 0
        self._peak_concurrency: int = 0

    def pre_init_operators(self, names: list[str]):
        """Pre-register all operator names so they appear in stats from startup."""
        with self._lock:
            for name in names:
                if name not in self._ops:
                    self._ops[name] = _OpStats()

    def _get_or_create(self, name: str) -> _OpStats:
        with self._lock:
            op = self._ops.get(name)
            if op is None:
                op = _OpStats()
                self._ops[name] = op
            return op

    def record_exec(self, name: str, duration_ns: int):
        self._get_or_create(name).record_exec(duration_ns)

    def record_skip(self, name: str):
        self._get_or_create(name).record_skip()

    def record_error(self, name: str, duration_ns: int):
        self._get_or_create(name).record_error(duration_ns)

    def record_run(self):
        with self._lock:
            self._run_count += 1

    def record_concurrency(self, n: int):
        with self._lock:
            if n > self._peak_concurrency:
                self._peak_concurrency = n

    def snapshot(self) -> dict[str, dict[str, Any]]:
        with self._lock:
            names = sorted(self._ops.keys())
        result: dict[str, dict[str, Any]] = {}
        for name in names:
            result[name] = self._get_or_create(name).snapshot()
        return result

    def scheduler_snapshot(self) -> dict[str, Any]:
        with self._lock:
            return {
                "run_count": self._run_count,
                "peak_concurrency": self._peak_concurrency,
            }
